internal/controllers/phase_ensure_subscription.go

```diff
@@ -0,0 +1,121 @@
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	addonsv1alpha1 "github.com/openshift/addon-operator/apis/addons/v1alpha1"
+)
+
+func (r *AddonReconciler) ensureSubscription(
+	ctx context.Context,
+	log logr.Logger,
+	addon *addonsv1alpha1.Addon,
+	catalogSource *operatorsv1alpha1.CatalogSource,
+) (
+	currentCSVKey client.ObjectKey,
+	requeue bool,
+	err error,
+) {
+	var commonInstallOptions addonsv1alpha1.AddonInstallOLMCommon
+	switch addon.Spec.Install.Type {
+	case addonsv1alpha1.OLMAllNamespaces:
+		commonInstallOptions = addon.Spec.Install.
+			OLMAllNamespaces.AddonInstallOLMCommon
+	case addonsv1alpha1.OLMOwnNamespace:
+		commonInstallOptions = addon.Spec.Install.
+			OLMOwnNamespace.AddonInstallOLMCommon
+	}
+
+	desiredSubscription := &operatorsv1alpha1.Subscription{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      addon.Name,
+			Namespace: commonInstallOptions.Namespace,
+		},
+		Spec: &operatorsv1alpha1.SubscriptionSpec{
+			CatalogSource:          catalogSource.Name,
+			CatalogSourceNamespace: catalogSource.Namespace,
+			Channel:                commonInstallOptions.Channel,
+			Package:                commonInstallOptions.PackageName,
+			// InstallPlanApproval is deliberately unmanaged
+			// API default is `Automatic`
+			// Legacy behavior of existing managed-tenants tooling is:
+			// All addons initially have to be installed with `Automatic`
+			// so that the very first InstallPlan succeedes
+			// but some addons want to take control of upgrades and thus
+			// change the Subscription.Spec.InstallPlanApproval value to `Manual`
+			// ATTENTION: When reconciling the subscription, we need to
+			// make sure to keep the current value of this field
+		},
+	}
+	addCommonLabels(desiredSubscription.Labels, addon)
+	if err := controllerutil.SetControllerReference(addon, desiredSubscription, r.Scheme); err != nil {
+		return client.ObjectKey{}, false, fmt.Errorf("setting controller reference: %w", err)
+	}
+
+	observedSubscription, err := r.reconcileSubscription(
+		ctx, desiredSubscription)
+	if err != nil {
+		return client.ObjectKey{}, false, fmt.Errorf("reconciling Subscription: %w", err)
+	}
+
+	if len(observedSubscription.Status.InstalledCSV) == 0 ||
+		len(observedSubscription.Status.CurrentCSV) == 0 {
+		log.Info("requeue", "reason", "csv not linked in subscription")
+		return client.ObjectKey{}, true, nil
+	}
+
+	installedCSVKey := client.ObjectKey{
+		Name:      observedSubscription.Status.InstalledCSV,
+		Namespace: commonInstallOptions.Namespace,
+	}
+	currentCSVKey = client.ObjectKey{
+		Name:      observedSubscription.Status.CurrentCSV,
+		Namespace: commonInstallOptions.Namespace,
+	}
+
+	changed := r.csvEventHandler.ReplaceMap(addon, installedCSVKey, currentCSVKey)
+	if changed {
+		log.Info("requeue", "reason", "csv-addon mapping changed")
```

Comment:
```suggestion
		// Mapping changes need to requeue, because we could have lost events before or during setting up the mapping, see csvEventHandler implementation for a longer description.
		log.Info("requeue", "reason", "csv-addon mapping changed")
```